brainframe_qt/util/oauth/cognito.py

```diff
@@ -0,0 +1,207 @@
+        if auth_response.state != self._state:
+            self.authentication_error.emit(self.InvalidStateError())
+            return
+
+        network_manager = QNetworkAccessManager(parent=self)
+        network_manager.finished.connect(self._on_token_reply)
+
+        # Surely there's a better way to do this
+        callback_url = QUrl.toPercentEncoding(self.callback_url.toString())
+        callback_url = bytes(callback_url).decode()
+
+        body = (
+            f"grant_type=authorization_code"
+            f"&client_id={self.client_id}"
+            f"&code_verifier={self._code_verifier}"
+            f"&code={auth_response.authorization_code}"
+            f"&redirect_uri={callback_url}"
+        )
+        request = QNetworkRequest(self.token_url)
+        request.setHeader(
+            request.ContentTypeHeader, "application/x-www-form-urlencoded"
+        )
+
+        network_manager.post(request, body.encode())
+
+    def _on_reply_handler_ready(self) -> None:
+        self.ready_to_authenticate.emit(self.authorization_url)
+
+    def _on_token_reply(self, reply: QNetworkReply) -> None:
+        """Called when the Access and Refresh Token completes"""
+        try:
+            response = json.loads(bytes(reply.readAll()))
+        except JSONDecodeError as exc:
+            exception = self.InvalidTokenResponse()
+            exception.__cause__ = exc
+
+            self.authentication_error.emit(exception)
+            return
+
+        try:
+            access_token = response["access_token"]
+            refresh_token = response["refresh_token"]
+        except KeyError as exc:
+            exception = self.InvalidTokenResponse()
+            exception.__cause__ = exc
+
+            self.authentication_error.emit(exception)
+            return
+
+        tokens = bf_codecs.CloudTokens(
+            access_token=access_token,
+            refresh_token=refresh_token,
+        )
+        self.authentication_successful.emit(tokens)
+
+    @staticmethod
+    def _make_pkce_code() -> Tuple[str, str]:
+        """Generates a PKCE code and returns a code challenge and code verifier. The
+        code challenge is provided to Cognito when the flow starts, and the code
+        verifier is provided when requesting the access and refresh tokens. These values
+        are used by Cognito to ensure that it's talking to the same client throughout
+        the entire flow.
+        """
+        alphanumeric = string.ascii_letters + string.digits
+        code_verifier = "".join(secrets.choice(alphanumeric) for _ in range(128))
+
+        code_challenge = hashlib.sha256(code_verifier.encode()).digest()
+        code_challenge = (base64.urlsafe_b64encode(code_challenge)
+                          .decode()
+                          .replace("=", ""))
+
+        return code_challenge, code_verifier
+
+    @staticmethod
+    def _make_state() -> str:
+        return str(uuid.uuid4())
+
+
+if __name__ == '__main__':
```

Comment:
I presume this is still helpful code you'd like to keep around?

Reply:
Yeah, I believe there are a few of these in the code base to make testing specific components easier. Seems pretty harmless to me.